mqtt/mqtt3/src/client/publish.rs

```diff
@@ -365,6 +365,11 @@ impl Default for State {
 pub struct PublishHandle(futures_channel::mpsc::Sender<PublishRequest>);
 
 impl PublishHandle {
+    /// Create a new `PublishHandle`
+    pub fn new(sender: futures_channel::mpsc::Sender<PublishRequest>) -> PublishHandle {
```

Comment:
Making things `Clone`able is fine with me, but I'm curious about this change. What use case do you have where you need to construct your own `PublishHandle`? If you use client in your agent, can you abstract that client with a trait and mock that instead of mocking parts of the client?

Reply:
Also, could [mockall](https://docs.rs/mockall/latest/mockall) help?

Reply:
We plan to do that in a future iteration -- in other words, we will create a trait that defines an interface for any MQTT client and then have one implementation that uses the iotedge client or another that uses a different crate. For now, our code is tightly coupled to the iotedge client (i.e. it uses the `PublishHandle` directly) and would require a significant refactor to do this. I need the changes in this PR because I introduced the following stop gap trait that allows us to increase code coverage significantly to help us prevent against regressions in the short term:

```
// Defines an abstract interface for an MQTT client.
// This interface builds on top of the Stream trait to define an interface that
// supports both receiving data (via the Stream interface/trait) and publishing data.
pub trait MqttClient: Stream + Unpin + Send {
    fn new() -> Result<Self, MqttManagerError>
    where
        Self: Sized;
    fn publish_handle(&self) -> Result<PublishHandle, MqttManagerError>;
}
```
While this trait is tightly coupled to the `PublishHandle` it allows us to create a `MockMqttClient` impl for testing our code without having to be in an edge container as the [azure-iot-mqtt Client](https://github.com/Azure/iotedge/blob/main/edge-modules/api-proxy-module/rust-sdk/azure-iot-mqtt/src/module.rs#L113-L117) requires. In our `MockMqtClient` is where we need to construct a `PublishHandle`.